src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs

```diff
@@ -108,7 +108,7 @@ public static OperationStatus EncodeToUtf8(ReadOnlySpan<byte> bytes, Span<byte>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static int GetMaxEncodedToUtf8Length(int length)
         {
-            if (length < 0 || length > MaximumEncodeLength)
+            if ((uint)length > MaximumEncodeLength)
```

Comment:
The case `length < 0` is covered due the cast to `uint`.

Reply:
This hides intent, though (that negative lengths are invalid).  And without `unchecked`, it would throw if the correct compiler/runtime flags are flipped.  
My naive guess is that the compiler has some built-in way to perform this particular optimization anyways; "positive-but-less-than" is a common category.

Reply:
`length` will just be reinterpreted as `uint`, so negative numbers are interpreted as very high (> int.MaxValue) numbers. `MaximumEncodeLength` is a constant (so no cast necessary) and on negative numbers the branch will be taken.

It's no compiler magic here, just interpretation of (raw-) values.

I don't think it hides intent, as used widely and as common trick to safe comparisons.

Reply:
>`length` will just be reinterpreted as `uint`, so negative numbers are interpreted as very high (> int.MaxValue) numbers.

Yeah.  [That's what causes the throw](https://dotnetfiddle.net/dKrY7J) (`checked` here is a way to force the [`-checked` compiler option](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/checked-compiler-option)) .

>I don't think it hides intent, as used widely and as common trick to safe comparisons.

The bare-bones form is demonstrably not always safe.
The equivalent behavior in C++ is officially **undefined** (although pretty much all implementations will behave as anticipated, for multiple reasons)

Reply:
@Clockwork-Muse, we do this optimization all over the place.